### PR TITLE
digitizing improvements

### PR DESCRIPTION
**Pull request by 3nids**
* saving does not add a point
* freeze rubber band while form is open

### CODE REVIEW

**Comment by m-kuhn** [comment]
Cool.
How does the "saving does not add point" work? Is there a different color used for the part which is not saved?

File: src/rubberbandmodel.cpp [hunk 1]
```diff
@@ -162,6 +163,9 @@ void RubberbandModel::setCurrentCoordinate( const QgsPoint& currentCoordinate )
   if ( mPointList.at( mCurrentCoordinateIndex ) == currentCoordinate )
     return;
 
+  if (mFreezed)
```
Comment (m-kuhn) [inline src/rubberbandmodel.cpp]: Is there a pre-commit hook defined? I think I once copied the one from QGIS.

And sidenote, according to my English dict, it's "freeze, froze, frozen" ;)

File: src/rubberbandmodel.cpp [hunk 2]
```diff
@@ -22,6 +22,7 @@ RubberbandModel::RubberbandModel( QObject* parent )
   , mCurrentCoordinateIndex( 0 )
   , mGeometryType( QgsWkbTypes::LineGeometry )
   , mLayer( nullptr )
+  , mFreezed ( false )
```
Comment (m-kuhn) [inline src/rubberbandmodel.cpp]: Header definition instead of here?

**Comment by m-kuhn** [comment]
@signedav can you also give that a review?

**Comment by qfield-fairy** [comment]
Uploaded test apks for [armv7](https://opengis.ch/download/qfield/ci-builds/qfield-dev-305-a720081b5fe5eeb22d2b385105037c5f706816ed-armv7.apk) and [x86](https://opengis.ch/download/qfield/ci-builds/qfield-dev-305-a720081b5fe5eeb22d2b385105037c5f706816ed-x86.apk)

**Comment by 3nids** [comment]
> How does the "saving does not add point" work? 
 the last point is removed before applying the geometry. 

> Is there a different color used for the part which is not saved?
not at the moment, I can add one

**Comment by m-kuhn** [comment]
> not at the moment, I can add one

I think it's quite important in terms of UX, that the last triangle is not part of what is being "pushed" to the layer.
Is this easily done? If not, better split this change into a separate pull request.

File: src/qml/DigitizingToolbar.qml [hunk 3]
```diff
@@ -72,6 +72,7 @@ Row {
     bgcolor: "#FFD600"
 
     onClicked: {
+      vertexRemoved()
```
Comment (signedav) [inline src/qml/DigitizingToolbar.qml]: I'm not sure about that. What is when I make:
- Start digitizing (pen-icon)
- Draw a line and press +
- Draw another line and then press the floppy (save)
Then the form opens but the last expected point (to have a triangle) is not set anymore.

Comment (signedav) [reply src/qml/DigitizingToolbar.qml]: I will precise myself. Wait one moment, then I show you in pictures.

Comment (signedav) [reply src/qml/DigitizingToolbar.qml]: When made (pen) draw line (+) draw line:
screenshot_2018-08-03-14-42-38
On the floppy here I expect it saves the third point (for a triangle)...

But this next example is improved in your version thanks to the freeze:
And here I made (pen) draw line (+) draw line (+) draw line (floppy) -> form opens:
screenshot_2018-08-03-14-27-30
This is the old version. Here it saves not a square (4 corners) when I press the final floppy but a pentagon (5 corners) what is not expected. So this is much better in your version.

Comment (3nids) [reply src/qml/DigitizingToolbar.qml]: indeed it needs to check when there is enough points to save, will fix it

**Comment by qfield-fairy** [comment]
Uploaded test apks for [armv7](https://opengis.ch/download/qfield/ci-builds/qfield-dev-305-51240bbbdb4334d712f1564bd0e05c6eb194dc0e-armv7.apk) and [x86](https://opengis.ch/download/qfield/ci-builds/qfield-dev-305-51240bbbdb4334d712f1564bd0e05c6eb194dc0e-x86.apk)

**Comment by signedav** [comment]
I tested it again and it works like intuitively expected.
Good work @3nids I approve.

**Comment by qfield-fairy** [comment]
Uploaded test apks for [armv7](https://opengis.ch/download/qfield/ci-builds/qfield-dev-305-42bdbc8740082c2fcd6c3306d754c538b9f957b9-armv7.apk) and [x86](https://opengis.ch/download/qfield/ci-builds/qfield-dev-305-42bdbc8740082c2fcd6c3306d754c538b9f957b9-x86.apk)

**Comment by 3nids** [comment]
@m-kuhn I added the second rubberband, this works well.
I looked to use a dashed line....quickly and found nothing.
https://stackoverflow.com/a/18447246/1548052

Any idea?

**Comment by signedav** [comment]
I confirm the crash with points is fixed. Thanks.

**Comment by qfield-fairy** [comment]
Uploaded test apks for [armv7](https://opengis.ch/download/qfield/ci-builds/qfield-dev-305-9dcacb0cbf261509011d5e1f41dd55691c41a78a-armv7.apk) and [x86](https://opengis.ch/download/qfield/ci-builds/qfield-dev-305-9dcacb0cbf261509011d5e1f41dd55691c41a78a-x86.apk)

**Comment by qfield-fairy** [comment]
Uploaded test apks for [armv7](https://opengis.ch/download/qfield/ci-builds/qfield-dev-305-adc32d9186916c649f2340482ae87e23a9e202e4-armv7.apk) and [x86](https://opengis.ch/download/qfield/ci-builds/qfield-dev-305-adc32d9186916c649f2340482ae87e23a9e202e4-x86.apk)

**Comment by signedav** [comment]
First it looks nice. I like it with lower opacity, like on digitizing.

Not sure if I understand it correctly, but this improvement is "only" that there is not moved a lost point. The stuff with lower opacity on open form shouldn't be saved in the end, correct?

Then there is a small issue. You make a square (pen, +, +, +, save) - when the form opens, you can move the last **saved** point by "++ instead of the point that was not saved by "save". Means, you have a fix triangle and are moving the cross-hair, like you would finish the square. In the end (on save of form), the square is saved like it should. So it's only wrong what is displayed...

Tell me if you need more info what it's about or if I should check in the code where the issue could be.

**Comment by 3nids** [comment]
I reverted the last commit which brought more troubles than it solved.

Now the only issue I see is if you "escape" the feature form while digitizing, you still have the `+` button but doesn't work. Altough, I think this issue is already present in master.

File: src/rubberbandmodel.cpp [hunk 4]
```diff
@@ -233,3 +239,14 @@ void RubberbandModel::setVectorLayer( QgsVectorLayer* layer )
...
+
+
+bool RubberbandModel::frozen() const
+{
+  return mFrozen;
+}
+
+void RubberbandModel::setFrozen(const bool &frozen )
+{
+  mFrozen = frozen;
```
Comment (m-kuhn) [inline src/rubberbandmodel.cpp]: emit?

File: src/rubberbandmodel.cpp [hunk 5]
```diff
@@ -186,6 +191,7 @@ void RubberbandModel::removeVertex()
 void RubberbandModel::reset()
 {
   removeVertices( 0, mPointList.size() - 1 );
+  mFrozen = false;
```
Comment (m-kuhn) [inline src/rubberbandmodel.cpp]: emit

File: src/sgrubberband.cpp [hunk 6]
```diff
@@ -8,6 +8,9 @@ SGRubberband::SGRubberband( const QVector<QgsPoint>& points, QgsWkbTypes::Geomet
 {
   mMaterial.setColor( color );
 
+  if ( points.count() == 0 )
```
Comment (m-kuhn) [inline src/sgrubberband.cpp]: empty()

**Comment by qfield-fairy** [comment]
Uploaded test apks for [armv7](https://opengis.ch/download/qfield/ci-builds/qfield-dev-305-656845f82eb7fd876d4c5c7bea815a090c2bc0d8-armv7.apk) and [x86](https://opengis.ch/download/qfield/ci-builds/qfield-dev-305-656845f82eb7fd876d4c5c7bea815a090c2bc0d8-x86.apk)

File: src/rubberbandmodel.cpp [hunk 7]
```diff
@@ -41,13 +41,15 @@ QVector<QgsPoint> RubberbandModel::vertices() const
   return mPointList;
 }
 
-QVector<QgsPoint> RubberbandModel::flatVertices() const
+QVector<QgsPoint> RubberbandModel::flatVertices(const bool &skipCurrentPoint) const
```
Comment (m-kuhn) [inline src/rubberbandmodel.cpp]: Copy trivial data structures by value

**Comment by 3nids** [comment]
escaping form should be solved by #309

File: src/rubberbandmodel.cpp (hunk 4 above)
Comment (3nids) [reply src/rubberbandmodel.cpp]: if there is no use of the signal, shall I rather remove it (the signal)?

Comment (m-kuhn) [reply src/rubberbandmodel.cpp]: It's in the property. And you can bind qml properties to this.

Comment (m-kuhn) [reply src/rubberbandmodel.cpp]: And better add a

```
if ( mFrozen == frozen )
  return;
```

It avoids deadlocks... Just in case.

Comment (3nids) [reply src/rubberbandmodel.cpp]: fixed

File: src/rubberbandmodel.cpp (hunk 5 above)
Comment (3nids) [reply src/rubberbandmodel.cpp]: fixed

**Comment by qfield-fairy** [comment]
Uploaded test apks for [armv7](https://opengis.ch/download/qfield/ci-builds/qfield-dev-305-44ad85821682a5e2215d644662d4a13576fe7704-armv7.apk) and [x86](https://opengis.ch/download/qfield/ci-builds/qfield-dev-305-44ad85821682a5e2215d644662d4a13576fe7704-x86.apk)

**Pull request merged**